### PR TITLE
iox-#456 fixed and tested variant queue chunk lost bug

### LINKED ISSUE #456: Chunks are lost forever when having an overflow in a variant queue of type FIFO
## Required information

**Operating system:**
Gentoo

**Compiler version:**
GCC 9.3.0

**Observed result or behaviour:**
When using a variant queue with type FIFO in a ChunkQueue, chunks are lost forever on overflow (memory leak)

**Expected result or behaviour:**
When a ChunkQueue has an overflow, the overflow is indicated with `hasOverflown()`. The chunks are lost for the subscriber but shouldn't be lost forever in the mempool 

**Conditions where it occurred / Performed steps:**
We are not using FIFOs in the ChunkQueues by default. Only in an integration test. Setup a ChunkQueue of type FIFO_SingleProducer_SingleConsumer. Push more chunks in the queue than the queue capacity. Delete all objects, check the number of free chunks in the mempool

### PR DESCRIPTION
**Pull request by elfenpiff**
Signed-off-by: Christian Eltzschig <me@elchris.org>

## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [ ] All checks have passed
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [ ] All checkboxes in the PR checklist are checked or crossed out
1. [ ] Merge

## References

- Closes #456

### CODE REVIEW

File: iceoryx_posh/test/moduletests/test_popo_chunk_queue.cpp [hunk 1]
```diff
@@ -247,13 +246,22 @@ TYPED_TEST(ChunkQueueFiFo_test, PushFull)
...
-    EXPECT_THAT(ret.get_error(), Eq(iox::popo::ChunkQueueError::QUEUE_OVERFLOW));
-    EXPECT_THAT(this->m_popper.empty(), Eq(false));
+
+    {
+        auto chunk = this->allocateChunk();
+        this->m_pusher.push(chunk);
+        EXPECT_TRUE(this->m_popper.hasOverflown());
+    }
+    
+    // get al the chunks in the queue
```
Comment (budrus) [inline iceoryx_posh/test/moduletests/test_popo_chunk_queue.cpp:258]: typo all

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl [hunk 2]
```diff
@@ -43,43 +43,33 @@ ChunkQueuePusher<ChunkQueueDataType>::getMembers() noexcept
...
-inline cxx::expected<ChunkQueueError> ChunkQueuePusher<ChunkQueueDataType>::tryPush(mepoo::SharedChunk chunk) noexcept
+inline void ChunkQueuePusher<ChunkQueueDataType>::push(mepoo::SharedChunk chunk) noexcept
 {
     ChunkTuple chunkTupleIn(chunk.releaseWithRelativePtr());
 
     auto pushRet = getMembers()->m_queue.push(chunkTupleIn);
 
-    if (pushRet.has_error())
+
+    // drop the chunk if one is returned by on overflow
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl:53]: typo. by an overflow

Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl:53]: fixed

File: iceoryx_posh/test/moduletests/test_popo_chunk_queue.cpp (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_posh/test/moduletests/test_popo_chunk_queue.cpp:258]: fixed

**Review by budrus — APPROVED** [review]

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/variant_queue.inl [hunk 3]
```diff
@@ -50,57 +50,48 @@ inline VariantQueue<ValueType, Capacity>::VariantQueue(const VariantQueueTypes t
...
 {
-    expected<optional<ValueType>, VariantQueueError> ret = error<VariantQueueError>(VariantQueueError::InternalError);
+    optional<ValueType> ret = cxx::nullopt_t();
 
     switch (m_type)
     {
     case VariantQueueTypes::FiFo_SingleProducerSingleConsumer:
     {
-        auto hasSuccess =
+        auto hadSpace =
```
Comment (mossmaurice) [inline iceoryx_utils/include/iceoryx_utils/internal/cxx/variant_queue.inl:61]: I'd prefer to have something like `pushWasSuccessful`. No strong feeling on this

**Review by elBoberido — APPROVED** [review]

**Review by mossmaurice — APPROVED** [review]

**Pull request merged**